modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/RowHandler.java

```diff
@@ -40,6 +40,9 @@
     /** */
     int columnCount(Row row);
 
+    /** */
+    Object[] getColumns(Row row);
```

Comment:
It's not only a shortcut but there is also a lot of overhead when we copy each field of row one by one. And GroupKey class still needs to be changed (at least `fieldsCount` method required). I think the current change doesn't break any abstraction but produces less garbage and works faster, but I can rewrite it to use one-by-one fields copy methods if you are not agreed. WDYT?

Reply:
Please don't be fooled by the fact we have only one row representation -- array of object. You should treat a row as an abstract row. I don't like the idea to extend the interface that way because this change is too implementation specific (in terms of reducing garbage).

> And GroupKey class still needs to be changed

I'm OK with this. Moreover I would rather generify a GroupKey. It should be a wrapper for the Row, and there is nothing wrong to return underlying row. But perhaps it would be better to do under separate ticket.

Reply:
> Please don't be fooled by the fact we have only one row representation -- array of object.

There can be any row representation. As far as `RowHandler` can return `Object` by the index of column and count of columns it can return `Object[]` - the array of columns values as well. The `getColumns()` method of `RowHandler` doesn't mean "give me internals of row", it means "make an array from columns values", but in some cases (`ArrayRowHandler`) it can be optimized and return just internals.

Reply:
>  As far as RowHandler can return Object by the index of column and count of columns it can return Object[]

should we provided similar methods for `List<Object>` and other collections then? And what about `BinaryObject getColumns(Row row)`?